htmd/parameterization/detect.py

```diff
@@ -0,0 +1,366 @@
+    """
+    Generate a molecular graph for methyl group
+    """
+
+    methyl = nx.Graph()
+    methyl.add_node(0, element='C')
+    methyl.add_node(1, element='H')
+    methyl.add_node(2, element='H')
+    methyl.add_node(3, element='H')
+    methyl.add_edge(0 ,1)
+    methyl.add_edge(0, 2)
+    methyl.add_edge(0, 3)
+
+    return methyl
+
+def detectParameterizableCores(graph):
+    """
+    Detect parametrizable dihedral angle cores (central atom pairs)
+
+    The cores are detected by looking for bridges (bonds with divide the molecule into two parts) in a molecular graph.
+    Terminal cores are skipped.
+    """
+
+    methyl = getMethylGraph()
+
+    all_core_sides = []
+    for core in list(nx.bridges(graph)):
+
+        # Get side graphs of the core
+        graph.remove_edge(*core)
+        sideGraphs = list(nx.connected_component_subgraphs(graph))
+        graph.add_edge(*core)
+
+        # Skip terminal bridges, which cannot form dihedral angles
+        if len(sideGraphs[0]) == 1 or len(sideGraphs[1]) == 1:
+            continue
+
+        # Swap the side graphs to match the order of the core
+        sideGraphs = sideGraphs[::-1] if core[0] in sideGraphs[1] else sideGraphs
+        assert core[0] in sideGraphs[0] and core[1] in sideGraphs[1]
+
+        # Skip if a side graph is a methyl group
+        if checkIsomorphism(sideGraphs[0], methyl) or checkIsomorphism(sideGraphs[1], methyl):
+            continue
+
+        # Skip if core contains C with sp hybridization
+        if graph.nodes[core[0]]['element'] == 'C' and graph.degree(core[0]) == 2:
+            continue
+        if graph.nodes[core[1]]['element'] == 'C' and graph.degree(core[1]) == 2:
+            continue
+
+        all_core_sides.append((core, sideGraphs))
+
+    return all_core_sides
+
+def weighted_closeness_centrality(graph, node, weight=None):
+    '''
+    Weighted closeness centrality
+
+    Identical to networkx.closeness_centrality, except the shorted path lengths are weighted by a node attribute.
+    '''
+
+    lengths = nx.shortest_path_length(graph, source=node)
+    del lengths[node]
+    weights = {node_: graph.nodes[node_][weight] for node_ in lengths} if weight else {node_: 1 for node_ in lengths}
+    centrality = sum(weights.values())/sum([lengths[node_]*weights[node_] for node_ in lengths])
+
+    return centrality
+
+def chooseTerminals(graph, centre, sideGraph):
+    """
+    Choose dihedral angle terminals (outer atoms)
+
+    The terminals are chosen by:
+    1. Largest closeness centrality
+    2. Largest atomic number weighted closeness centrality
+    """
+
+    terminals = list(sideGraph.neighbors(centre))
+
+    # Get a subgraph for each terminal
+    sideGraph = sideGraph.copy()
+    sideGraph.remove_node(centre)
+    terminalGraphs = list(nx.connected_component_subgraphs(sideGraph))
+    terminalGraphs = [[terminalGraph for terminalGraph in terminalGraphs if terminal in terminalGraph.nodes][0] for terminal in terminals]
```

Comment:
Rewrite this please as a double loop. Nested list comprehensions are hard to read

Reply:
Rewrote as a single-loop list comprehension... :D